opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/Spea2.java

```diff
@@ -406,10 +382,48 @@ public int compare(Spea2IndividualSet o1, Spea2IndividualSet o2) {
 		return lames;
 	}
 
+	/**
+	 * Helper function to preorder the candidates for getNearest() operation
+	 * 
+	 * @param candidates
+	 *            the candidate IndividualSets
+	 * @return the preordered IndividualSets
+	 */
+	protected Map<Spea2IndividualSet, List<Spea2IndividualSet>> getNearestPreOrder(
+			Collection<Spea2IndividualSet> candidates) {
+		Map<Spea2IndividualSet, List<Spea2IndividualSet>> orderedLists = new LinkedHashMap<Spea2IndividualSet, List<Spea2IndividualSet>>();
+
+		for (Spea2IndividualSet w0 : candidates) {
+			List<Spea2IndividualSet> list = new ArrayList<Spea2IndividualSet>();
+			for (Spea2IndividualSet w1 : candidates) {
+				if (w0 != w1) {
+					w1.setNextDistance(distance(w0, w1));
+					list.add(w1);
+				}
+			}
+			Collections.sort(list, new Comparator<Spea2IndividualSet>() {
+				@Override
+				public int compare(Spea2IndividualSet o1, Spea2IndividualSet o2) {
+					double v = o1.getNextDistance() - o2.getNextDistance();
+					if (v < 0) {
+						return -1;
+					} else if (v > 0) {
+						return 1;
+					} else {
+						return 0;
+					}
+				}
+
+			});
+			orderedLists.put(w0, list);
+		}
+		return orderedLists;
+	}
+
 	protected double getMinDistance(Spea2IndividualSet w0) {
 		double min = Double.MAX_VALUE;
 		for (Spea2IndividualSet w1 : individualSets) {
-			if (w0 != w1) {
+			if (w0 == w1) {
```

Comment:
Has this been a logic error? Then, we should outline this explicitly in the release notes of the next release and also make this bug fix release quite soon.

Reply:
That was an important catch! Thanks a lot. I made that mistake reverting back from equals(). Corrected.